Fix decorators to allow users to filter test cases by number of GPUs

kmaehashi opened the pull request:
`pytest` does not support filtering test cases based on parameters of decorators.
(`pytest -m MARKEXPR` only checks [existence of marks and does not evaluate parameters](https://github.com/pytest-dev/pytest/blob/3.2.3/_pytest/mark.py#L167-L199))

This means that filter based on number of GPUs (as documented in the [Contribution Guide](https://docs.chainer.org/en/stable/contribution.html#how-to-run-tests)):

```
$ nosetests path/to/gpu/test.py --eval-attr='gpu<2'
```

no longer work.

This PR fixes `multi_gpu` decorator so that `CHAINER_TESTING_GPU_LIMIT` environment variable can be used to limit number of GPUs used.

The above command needs to be rewritten as:

```
$ export CHAINER_TESTING_GPU_LIMIT=1
$ pytest path/to/gpu/test.py
```

takagi commented:
Would you rebase the master so that Travis passes?

takagi commented:
LGTM! I will merge this after Travis will pass.

takagi commented:
Would you note about `CHAINER_TESTING_GPU_LIMIT ` environment variable on the reference? https://github.com/chainer/chainer/blob/master/docs/source/reference/environment.rst

kmaehashi commented:
Thank you for review.

I was not aware of the reference. Added ``CHAINER_TESTING_GPU_LIMIT` to it.

BTW I found many of environment variables are undocumented. I'll fix this in another PR.

* CHAINER_PYTHON_350_FORCE
* CHAINER_DEBUG
* CHAINER_KEEP_GRAPH_ON_REPORT
* CHAINER_USE_CUDNN
* CHAINER_DATASET_ROOT

kmaehashi commented:
I found `CHAINER_TEST_RANDOM_NONDETERMINISTIC` is used here.
https://github.com/chainer/chainer/blob/v3.0.0/chainer/testing/random.py#L62-L63

Maybe the name should be `CHAINER_TEST_GPU_LIMIT`  instead of `CHAINER_TESTING_GPU_LIMIT`.

kmaehashi commented:
Renamed the environment variable name.

I also updated the reference in another PR to avoid conflict.  https://github.com/chainer/chainer/pull/3662/files

takagi commented:
Merged!

kmaehashi commented:
Ah, this broke AppVeyor test. I'll fix
